sql-composer/src/parser.rs

```diff
@@ -586,158 +576,153 @@ pub fn bind_value_real(span: Span) -> IResult<Span, SerdeValue> {
     let (span, _) = multispace0(span)?;
     let (span, _) = check_bind_value_ending(span)?;
     let (span, _) = multispace0(span)?;
+
     Ok((span, SerdeValue(Value::F64(value))))
 }
 
-#[cfg(feature = "composer-serde")]
-named!(
-    check_bind_value_ending(Span) -> Span,
-    alt!(
-        eof!()           | // shortest first
-        peek!(tag!(")")) |
-        peek!(tag!("]")) |
-        peek!(tag!(","))
-    )
+named!(ending(Span) -> Span,
+    eof!()
 );
 
 #[cfg(feature = "composer-serde")]
-named!(
-    bind_value(Span) -> (SerdeValue),
-    do_parse!(
-        value: complete!(
-            alt!(
-                do_parse!(t: bind_value_text >> (t))    |
-                do_parse!(i: bind_value_integer >> (i)) |  // must parse int before real
-                do_parse!(r: bind_value_real >> (r))
-            )
-        ) >>
-        ({
-            value
-        })
-    )
-);
+pub fn check_bind_value_ending(span: Span) -> IResult<Span, Span> {
+    let (span, _) = alt((ending, peek(tag(")")), peek(tag("]")), peek(tag(","))))(span)?;
+
+    Ok((span, span))
+}
 
 #[cfg(feature = "composer-serde")]
-named!(
-    bind_value_set(Span) -> Vec<SerdeValue>,
-    do_parse!(
-        start: opt!(complete!(alt!(tag!("[") | tag!("(")))) >>
-        list: fold_many1!(
-            do_parse!(
-                value: bind_value >>
-                multispace0 >>
-                opt!(complete!(tag!(","))) >>
-                multispace0 >>
-                (value)
-            ),
-            vec![], |mut acc: Vec<SerdeValue>, item: SerdeValue| {
-                acc.push(item);
-                acc
-            }) >>
-        end: opt!(complete!(alt!(tag!("]") | tag!(")")))) >>
-        ({
-            if let Some(s) = start {
-                if let Some(e) = end {
-                    if s.fragment == "[" && e.fragment != "]" {
-                        panic!("bind_value_set: no corresponding '[' for ']'");
-                    }
-                    else if s.fragment == "(" && e.fragment != ")" {
-                        panic!("bind_value_set: no corresponding ')' for '('");
-                    }
-                }
-                else {
-                    panic!("bind_value_set: no matching end found for start: {:?}", s);
-                }
+pub fn bind_value(span: Span) -> IResult<Span, SerdeValue> {
+    let (span, value) = alt((
+            bind_value_text,
+            bind_value_integer,
+            bind_value_real
+    ))(span)?;
+
+    Ok((span, value))
+}
+
+#[cfg(feature = "composer-serde")]
+pub fn bind_value_separated(span: Span) -> IResult<Span, SerdeValue> {
+    let (span, value) = bind_value(span)?;
+    let (span, _) = multispace0(span)?;
+    let (span, _) = opt(tag(","))(span)?;
+    let (span, _) = multispace0(span)?;
```

Comment:
Yeah, that name is bound to cause confusion. What about `comma_separator`?

Reply:
In my local source, I modified line 1485 in `test_bind_value_set` to remove an inner comma, and unfortunately it still parsed.

```
    let input = "['a', 'aa' 'aaa']";  //removed comma between 2nd and 3rd item
```

`"[ 'a', 'aa' 'aaa' ]"` returned `["a", "aa", "aaa"]`

Reply:
Also, we have a test for `test_bind_value_set`, but not for the sub-pieces like `bind_value_separated`

I'd like to have (small, at a minimum) tests for each sub parser, to show at least the intended happy-case.  This type of "see, it works, it goes from a => b" might fit well into the doc tests?  And leave the unit-tests for more expansive testing.

Reply:
Rules for commas: 
1. only allowed inside containers -- either brackets or parenthesis. 
1. outside containers, only a single value is allowed, so no commas.
1. trailing comma is not allowed

Does that sound about right for sql?

This would mean that we don't allow optional bracketing.  We allow an alt between matching a single item or between matching a set of items.  Something like:
```
    alt(bind_value, bind_value_set_non_optional_container_and_no_trailing_comma)(span)?
```

Does that `bind_value` need to be `bind_single_value_as_array` ?

Reply:
> Rules for commas:
> 
>     1. only allowed inside containers -- either brackets or parenthesis.

>     2. outside containers, only a single value is allowed, so no commas.
> 
>     3. trailing comma is not allowed
> 
> Does that sound about right for sql?
> 
Yes, these all sound correct. I need to think more on the rest of this question.